# #+TITLE: Digital Biology -- BIOL647
#+AUTHOR: Rodolfo Aramayo
#+EMAIL: raramayo@tamu.edu
#+STARTUP: align
* *Fundamentals of Transcriptome Mapping and Analysis* 
** *Outline of The Tuxedo Pipeline*
   | [[./00Data/T15Data/RNASeqFig01.jpg]] |
** *Mapping RNASeq reads to the genome*
   | *Outline*                     |
   | [[./00Data/T15Data/mapping_outline.png]] |
   To make sense of the data, we need to first figure out where the
   sequences originated from in the genome, so we can then determine to
   which genes they belong. When a reference genome for the organism is
   available, this process is known as aligning or "mapping" the reads
   to the reference. This is equivalent to solving a jigsaw puzzle, but
   unfortunately, not all pieces are unique.
   
   With eukaryotic transcriptomes most reads originate from processed
   mRNAs lacking introns:
   | [[./00Data/T15Data/rna-seq-reads.png]] |
   
   Therefore they cannot be simply mapped back to the genome as we
   normally do for DNA data. Spliced-awared mappers have been developed
   to efficiently map transcript-derived reads against a reference
   genome:
   | [[./00Data/T15Data/splice_aware_alignment.png]] |
   | *Outline*                            |
   | [[./00Data/T15Data/RNASeqFig04.jpg]]            |
   | *Split Alignment*                    |
   | [[./00Data/T15Data/RNASeqFig05.jpg]]            |
   
   To further optimize and speed up spliced read alignment, [[https://www.nature.com/articles/s41587-019-0201-4][HISAT2]] was
   developed. It uses a hierarchical graph FM (HGFM) index,
   representing the entire genome and eventual variants, together with
   overlapping local indexes (each spanning ~57 kb) that collectively
   cover the genome and its variants. This allows to find initial seed
   locations for potential read alignments in the genome using global
   index and to rapidly refine these alignments using a corresponding
   local index:
   | [[./00Data/T15Data/hisat.png]] |
   A part of the read (blue arrow) is first mapped to the genome using
   the global FM index. HISAT2 then tries to extend the alignment
   directly utilizing the genome sequence (violet arrow). In (a) it
   succeeds and this read is aligned as it completely resides within
   an exon. In (b) the extension hits a mismatch. Now HISAT2 takes
   advantage of the local FM index overlapping this location to find
   the appropriate mapping for the remainder of this read (green
   arrow). The (c) shows a combination these two strategies: the
   beginning of the read is mapped using global FM index (blue arrow),
   extended until it reaches the end of the exon (violet arrow),
   mapped using local FM index (green arrow) and extended again
   (violet arrow).
** *Counting the number of reads per annotated gene*
To compare the expression of single genes between different conditions
(e.g. with or without PS depletion), an essential first step is to
quantify the number of reads per gene, or more specifically the number
of reads mapping to the exons of each gene.
| [[./00Data/T15Data/gene_counting.png]] |
# + How many reads are found for the different exons?
# + How many reads are found for the different genes?
Two main tools could be used for that:
+ [[https://htseq.readthedocs.io/en/release_0.9.1/count.html][HTSeqCount]]
+ [[http://subread.sourceforge.net/][Subread - featureCounts]]
FeatureCounts is considerably faster and requires far less
computational resources.  In principle, the counting of reads
overlapping with genomic features is a fairly simple task. But there
are some details that need to be given to featureCounts, e.g. the
strandness.
*** *Estimation of the strandness*
RNAs that are typically targeted in RNA-Seq experiments are single
stranded (e.g., mRNAs) and thus have polarity (5' and 3' ends that are
functionally distinct). During a typical RNA-Seq experiment the
information about strandness is lost after both strands of cDNA are
synthesized, size selected, and converted into a sequencing
library. However, this information can be quite useful for the read
counting step, especially for reads located on the overlap of 2 genes
that are on different strands.
| [[./00Data/T15Data/strandness_why.png]] |
Some library preparation protocols create so called stranded RNA-Seq
libraries that preserve the strand information (an excellent overview
in [[https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3005310/][here]]. In practice, with Illumina paired-end RNA-Seq protocols you
are unlikely to encounter many of these possibilities. You will most
likely deal with either:
+ Unstranded RNA-Seq data
+ Stranded RNA-Seq data generated by the use of specialized RNA
  isolation kits during sample preparation
| *More details about strandness* |
| [[./00Data/T15Data/dna_rna.png]]           |
The implication of stranded RNA-Seq is that you can distinguish
whether the reads are derived from forward or reverse-encoded
transcripts:
| [[./00Data/T15Data/stranded_result.png]] |
Depending on the approach, and whether one performs single-end or
paired-end sequencing, there are multiple possibilities on how to
interpret the results of the mapping of these reads to the genome:
| [[./00Data/T15Data/rnaseq_library_type.png]] |
This information should be provided with your FASTQ files, ask your
sequencing facility! If not, try to find it on the site where you
downloaded the data or in the corresponding publication.

Another option is to estimate these parameters with a tool called
*Infer Experiment* from the [[https://www.ncbi.nlm.nih.gov/pubmed/22743226][RSeQC]] tool suite. This tool takes the BAM
files from the mapping, selects a subsample of the reads and compares
their genome coordinates and strands with those of the reference gene
model (from an annotation file). Based on the strand of the genes, it
can gauge whether sequencing is strand-specific, and if so, how reads
are stranded (forward or reverse):
| [[./00Data/T15Data/strandness_cases.png]] |
Infer Experiment tool generates one file with information on:
  + Paired-end or single-end library
  + Fraction of reads failed to determine
  + 2 lines
    + For single-end
      + Fraction of reads explained by "++,--": the fraction of reads that assigned to forward strand
      + Fraction of reads explained by "+-,-+": the fraction of reads that assigned to reverse strand
    + For paired-end
      + Fraction of reads explained by "1++,1--,2+-,2-+": the fraction of reads that assigned to forward strand
      + Fraction of reads explained by "1+-,1-+,2++,2--": the fraction of reads that assigned to reverse strand
If the two "Fraction of reads explained by" numbers are close to each
other, we conclude that the library is not a strand-specific dataset
(or unstranded).

As it is sometimes quite difficult to find out which settings
correspond to those of other programs, the following table might be
helpful to identify the library type:

|----------------------+-------------------------------+------------------+--------------------+-------------+----------------|
| Library type         | Infer Experiment              | TopHat           | HISAT2             | HTSeq-count | featureCounts  |
|----------------------+-------------------------------+------------------+--------------------+-------------+----------------|
| Paired-End (PE) - SF | 1 + + , 1 - - , 2 + - , 2 - + | FR Second Strand | Second Strand F/FR | yes         | Forward (1)    |
| PE - SR              | 1 + - , 1 - + , 2 + + , 2 - - | FR First Strand  | First Strand R/RF  | reverse     | Reverse (2)    |
| Single-End (SE) - SF | + + , - -                     | FR Second Strand | Second Strand F/FR | yes         | Forward (1)    |
| SE - SR              | + - , - +                     | FR First Strand  | First Strand R/RF  | reverse     | Reverse (2)    |
| PE, SE - U           | undecided                     | FR Unstranded    | default            | no          | Unstranded (0) |
|----------------------+-------------------------------+------------------+--------------------+-------------+----------------|

** *Mapping Complexity Considerations*
| *Complexity*                |
| [[./00Data/T15Data/RNASeqFig03.jpg]]   |
| *Identifying Fusion Events* |
| [[./00Data/T15Data/RNASeqFig06.jpg]]   |
** *Tuxedo Pipeline*
| *Outline*                 |
| [[./00Data/T15Data/RNASeqFig07.jpg]] |
------------
| *Comparing Conditions*    |
| [[./00Data/T15Data/RNASeqFig08.jpg]] |
------------
| *Analysis*                |
| [[./00Data/T15Data/RNASeqFig09.jpg]] |
------------
| [[./00Data/T15Data/RNASeqFig10.jpg]] |
------------
| [[./00Data/T15Data/RNASeqFig11.jpg]] |
------------
| [[./00Data/T15Data/RNASeqFig12.jpg]] |
------------
| [[./00Data/T15Data/RNASeqFig13.jpg]] |
*** *Tuxedo Pipeline Information*
|---------------+----------------------+---------------------------------------------------------------------------------------------------------|
| Program       | Manual               | Publications                                                                                            |
|---------------+----------------------+---------------------------------------------------------------------------------------------------------|
| TopHat        | [[http://ccb.jhu.edu/software/tophat/manual.shtml][TopHat2_Manual]]       | [[http://bioinformatics.oxfordjournals.org/content/25/9/1105.abstract][TopHat: discovering splice junctions with RNA-Seq]]                                                       |
|               |                      | [[http://genomebiology.com/2013/14/4/R36/abstract][TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions]] |
|---------------+----------------------+---------------------------------------------------------------------------------------------------------|
| TopHat-Fusion | [[http://ccb.jhu.edu/software/tophat/fusion_manual.html][TopHat-Fusion_Manual]] | [[http://genomebiology.com/2011/12/8/R72/abstract][TopHat-Fusion: an algorithm for discovery of novel fusion transcripts]]                                   |
|---------------+----------------------+---------------------------------------------------------------------------------------------------------|
| Cufflinks     | [[http://cole-trapnell-lab.github.io/cufflinks/manual/][Cufflinks_Manual]]     | [[http://www.nature.com/nbt/journal/v28/n5/full/nbt.1621.html][Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform ...]]       |
|               |                      | [[http://genomebiology.com/2011/12/3/R22/abstract][Improving RNA-Seq expression estimates by correcting for fragment bias]]                                  |
|               |                      | [[http://bioinformatics.oxfordjournals.org/content/27/17/2325][Identification of novel transcripts in annotated genomes using RNA-Seq]]                                  |
|               |                      | [[http://www.nature.com/nbt/journal/v31/n1/full/nbt.2450.html][Differential analysis of gene regulation at transcript resolution with RNA-seq]]                          |
|---------------+----------------------+---------------------------------------------------------------------------------------------------------|
*** *Tuxedo Pipeline Tutorials*
|-------------------+-------------------------------------------------------------------------------------------------------|
| Journal           | Title                                                                                                 |
|-------------------+-------------------------------------------------------------------------------------------------------|
| Nature Protocols: | [[http://www.nature.com/nprot/journal/v7/n3/abs/nprot.2012.016.html][Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks]] |
|-------------------+-------------------------------------------------------------------------------------------------------|
** *The New Pipelines*
*** *HISAT2, StringTie and Ballgown*
| *Overview*                |
| [[./00Data/T15Data/RNASeqFig14.jpg]] |
------------
| *Logic*                   |
| [[./00Data/T15Data/RNASeqFig15.jpg]] |
------------
| *Classes of Reads*        |
| [[./00Data/T15Data/RNASeqFig16.jpg]] |
------------
| *HISAT is Fast*           |
| [[./00Data/T15Data/RNASeqFig17.jpg]] |
*** *HISAT2, StringTie and Ballgown Information*
|--------------+------------------------------------------------------------------------------------------------|
| *PIPELINE:*  | *HISAT2, StringTie and Ballgown*                                                               |
|--------------+------------------------------------------------------------------------------------------------|
| *HISAT2:*    | [[https://ccb.jhu.edu/software/hisat2/manual.shtml][Manual]]                                                                                         |
|              | [[http://www.nature.com/nmeth/journal/v12/n4/full/nmeth.3317.html][HISAT: a fast spliced aligner with low memory requirements]]                                     |
|--------------+------------------------------------------------------------------------------------------------|
| *StringTie:* | [[http://ccb.jhu.edu/software/stringtie/index.shtml?t=manual][Manual]]                                                                                         |
|              | [[https://www.nature.com/articles/nbt.3122][StringTie enables improved reconstruction of a transcriptome from RNA-seq reads]]                |
|              | [[http://www.nature.com/nprot/journal/v11/n9/full/nprot.2016.095.html][Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown]] |
|--------------+------------------------------------------------------------------------------------------------|
| *Ballgown:*  | [[https://bioconductor.org/packages/release/bioc/manuals/ballgown/man/ballgown.pdf][Manual]]                                                                                         |
|              | [[https://github.com/alyssafrazee/ballgown][Documentation]]                                                                                  |
|              | [[https://www.nature.com/articles/nbt.3172][Ballgown bridges the gap between transcriptome assembly and expression analysis]]                |
|--------------+------------------------------------------------------------------------------------------------|
*** *STAR, StringTie and DESeq2 Information*
|--------------+------------------------------------------------------------------------------------------------|
| *PIPELINE:*  | *STAR, StringTie and DESeq2*                                                                   |
|--------------+------------------------------------------------------------------------------------------------|
| *STAR:*      | [[https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf][Manual]]                                                                                         |
|              | [[https://academic.oup.com/bioinformatics/article-lookup/doi/10.1093/bioinformatics/bts635][STAR: ultrafast universal RNA-seq aligner]]                                                      |
|--------------+------------------------------------------------------------------------------------------------|
| *StringTie:* | [[http://ccb.jhu.edu/software/stringtie/index.shtml?t=manual][Manual]]                                                                                         |
|              | [[https://www.nature.com/articles/nbt.3122][StringTie enables improved reconstruction of a transcriptome from RNA-seq reads]]                |
|              | [[http://www.nature.com/nprot/journal/v11/n9/full/nprot.2016.095.html][Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown]] |
|--------------+------------------------------------------------------------------------------------------------|
| *DESeq2:*    | [[http://bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf][Manual]]                                                                                         |
|              | [[http://www.nature.com/nprot/journal/v8/n9/full/nprot.2013.099.html][Count-based differential expression analysis of RNA sequencing data using R and Bioconductor]]   |
|              | [[http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html][Analyzing RNA-seq data with DESeq2]]                                                             |
|--------------+------------------------------------------------------------------------------------------------|
*** *Kallisto and Sleuth and Salmon Information*
|-------------+-----------------------------------------------------------------------------|
| *PIPELINE:* | *Kallisto and Sleuth and Salmon*                                            |
|-------------+-----------------------------------------------------------------------------|
| *Kallisto:* | [[https://pachterlab.github.io/kallisto/about][About]]                                                                       |
|             | [[https://www.nature.com/articles/nbt.3519][Near-optimal probabilistic RNA-seq quantification]]                           |
|-------------+-----------------------------------------------------------------------------|
| *Sleuth:*   | [[https://pachterlab.github.io/sleuth/about][About]]                                                                       |
|             | [[http://www.nature.com/nmeth/journal/v14/n7/full/nmeth.4324.html][Differential analysis of RNA-seq incorporating quantification uncertainty]]   |
|             | [[https://liorpachter.wordpress.com/2015/08/17/a-sleuth-for-rna-seq/][A sleuth for RNA-Seq]]                                                        |
|             | [[https://github.com/pachterlab/sleuth][GITHUB]]                                                                      |
|-------------+-----------------------------------------------------------------------------|
| *Salmon*    | [[https://combine-lab.github.io/salmon/][About]]                                                                       |
|             | [[http://salmon.readthedocs.io/en/latest/index.html][Manual]]                                                                      |
|             | [[https://github.com/COMBINE-lab/salmon][GITHUB]]                                                                      |
|             | [[https://www.nature.com/nmeth/journal/v14/n4/abs/nmeth.4197.html][Salmon provides fast and bias-aware quantification of transcript expression]] |
|-------------+-----------------------------------------------------------------------------|
*** *Blog Wars*
|-------------+-------------------------------------------------------------------------------------|
| *Blog_Wars* | [[https://liorpachter.wordpress.com/2017/08/02/how-not-to-perform-a-differential-expression-analysis-or-science/][How not to perform a differential expression analysis (or science)]]                  |
|             | [[https://github.com/salmonteam/SalmonBlogResponse/blob/master/SalmonBlogResponse.md][Response to the blog post about Salmon and kallisto]]                                 |
|             | [[http://robpatro.com/blog/?p=248][Not-quite alignments: Salmon, Kallisto and Efficient Quantification of RNA-Seq data]] |
|-------------+-------------------------------------------------------------------------------------|
*** *Other ALigners*
|-------------------+--------------------------------------------------------------------------|
| *OTHER Aligners:* |                                                                          |
|-------------------+--------------------------------------------------------------------------|
| *GMAP:*           | [[http://research-pub.gene.com/gmap/][Info]]                                                                     |
|                   | [[https://academic.oup.com/bioinformatics/article/21/9/1859/409207][GMAP: a genomic mapping and alignment program for mRNA and EST sequences]] |
|                   | [[https://github.com/PacificBiosciences/cDNA_primer/wiki/Aligner-tutorial:-GMAP,-STAR,-BLAT,-and-BLASR][Aligner tutorial: GMAP, STAR, BLAT, and BLASR]]                            |
|-------------------+--------------------------------------------------------------------------|
** *Evaluating Results:*
*** [[https://www.youtube.com/watch?v=TTUrtCY2k-w][RPKM, FPKM and TPM]]
** *Final Considerations*
| [[./00Data/T15Data/RNASeqFig19.jpg]]                                 |
| Make sure to tune your parameters...Not just use defaults |
* *References*
   + This lesson has been developed using materials from various
     sources, that include, but are not restricted to training
     tutorials developed by the [[https://github.com/galaxyproject/training-material][Galaxy Project]] team. These are open
     access materials distributed under the terms of the Creative
     Commons Attribution license (CC BY 4.0), which permits
     unrestricted use, distribution, and reproduction in any medium,
     provided the original author and source are credited.
   + Other Bibliographic References are:
     + [[./00Data/T15Data/RNA_Paper01.pdf][Comprehensive comparative analysis of strand-specific RNA sequencing methods]]
     + [[./00Data/T15Data/RNA_Paper02.pdf][Transcript assembly and quantification by RNA-seq..]]
     + [[./00Data/T15Data/RNA_Paper03.pdf][Identification of novel transcripts in annotated genomes using...]]
     + [[./00Data/T15Data/RNA_Paper04.pdf][Differential analysis of gene regulation at transcript...]]
     + [[./00Data/T15Data/RNA_Paper05_TopHat.pdf][TopHat: discovering splice junctions with RNA-Seq]]
     + [[./00Data/T15Data/RNA_Paper06_TopHatFusion.pdf][TopHat-Fusion: an algorithm for discovery...]]
     + [[./00Data/T15Data/RNA_Paper07_TopHatCufflinks.pdf][Differential gene and transcript expression analysis...]]
     + [[./00Data/T15Data/RNA_Paper08_TopHat2.pdf][TopHat2: accurate alignment of transcriptomes...]]
     + [[./00Data/T15Data/RNA_Paper09_HISAT.pdf][HISAT: a fast spliced aligner with low memory...]]
     + [[./00Data/T15Data/RNA_Paper10_Stringtie.pdf][Stringtie enables improved reconstruction...]]
     + [[./00Data/T15Data/RNA_Paper11_Ballgown.pdf][Count-based differential expression analysis of RNA...]]
     + [[./00Data/T15Data/RNA_Paper12_HISAT_Stringtie_Ballgown.pdf][Transcript-level expression analysis of RNA-seq...]]
     + [[./00Data/T15Data/RNA_Paper13_STAR.pdf][STAR: ultrafast universal RNA-seq aligner...]]
     + [[./00Data/T15Data/RNA_Paper14_Kallisto.pdf][Near-optimal probabilistic RNA-Seq quantification...]]
     + [[./00Data/T15Data/RNA_Paper15_Sleuth.pdf][Differential analysis of RNA-Seq...]]
     + [[./00Data/T15Data/RNA_Paper16_Salmon.pdf][Salmon provides fast...]]
     + [[./00Data/T15Data/RNA_Paper17_GMAP.pdf][GMAP: a genomic mapping and alignment...]]
     + [[./00Data/T15Data/RNA_Paper18.pdf][Spatially resolved transcriptomics...]]
     + [[./00Data/T15Data/RNA_Paper19.pdf][Simulation-Based omprehensive benchmarking...]]
     + [[./00Data/T15Data/RNA_Paper20.pdf][A survey of best practices for RNA-seq data Analysis]]
# ============
# * *Gene Set Enrichment*
# 1. Motivation
#    + Is expression of genes in a gene set associated with experimental condition?
#      + E.g., Are there unusually many up-regulated genes in the gene set?
# 2. Many methods, a recent review is [[http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1002375][Kharti et al., 2012]]
#    + Over-representation analysis (ORA) – are differentially expressed (DE) genes in the set more common than expected?
#    + Functional class scoring (FCS) – summarize statistic of DE of genes in a set, and compare to null
#    + Pathway topology (PT) – include pathway knowledge in assessing DE of genes in a set
# 3. What is a gene set?
#    + Any /a priori/ classification of `genes’ into biologically relevant groups
#      + Members of same [[http://biochemical-pathways.com/#/map/1][biochemical pathway]] (e.g., Steroid Metabolism - Cholesterol Synthesis)
#      + Proteins expressed in identical cellular compartments (e.g., Mitochondria)
#      + Co-expressed under certain conditions (e.g., starvation)
#      + Targets of the same regulatory elements
#      + On the same cytogenic band
#    + Sets do not need to be
#      + Exhaustive
#      + Disjoint
# 4. Collections of gene sets
#    + [[http://geneontology.org/][Gene Ontology (GO) Annotation (GOA)]]
#      + CC Cellular Components
#      + BP Biological Processes
#      + MF Molecular Function
#    + [[http://software.broadinstitute.org/gsea/msigdb/index.jsp][Molecular Signatures Database -- MSigDB]]
#      + The Molecular Signature Database ispart of a larger project: [[http://software.broadinstitute.org/gsea/index.jsp][Gene Set Enrichment Analysis Project]]
#      + E.g., MSigDb
#        + c1 Positional gene sets – chromosome & cytogenic band
#        + c2 Curated Gene Sets from online pathway databases, publications in PubMed, and knowledge of domain experts
#        + c3 motif gene sets based on conserved cis-regulatory motifs from a comparative analysis of the human, mouse, rat, and dog genomes
#        + c4 computational gene sets defined by mining large collections of cancer-oriented microarray data
#        + c5 GO gene sets consist of genes annotated by the same GO terms
#        + c6 oncogenic signatures defined directly from microarray gene expression data from cancer gene perturbations
#        + c7 immunologic signatures defined directly from microarray gene expression data from immunologic studies
# 5. Pathways
#    + [[http://www.broadinstitute.org/gsea/msigdb/index.jsp][MSigDb]]
#    + [[http://www.genome.jp/kegg/pathway.html][KEGG]]
#    + [[http://www.reactome.org/][Reactome]]
#    + [[http://pantherdb.org/][PantherDB]]
# 6. Statistical approaches ([[http://marray.economia.unimi.it/2009/material/lectures/L8_Gene_Set_Testing.pdf][Based on a presentation by Simon Anders, CSAMA 2010]])
#    + Approach 1: [[http://bioconductor.org/packages/release/bioc/html/GOstats.html][hypergeometric tests]]
#      + Steps
#        + Classify each gene as ‘differentially expressed’ DE or not, e.g., based on P < 0.05
#        + Are DE genes in the set more common than DE genes not in the set?
#    + Approach 2: [[http://www.pnas.org/content/102/43/15545.long][enrichment score]]
#      + Steps
#        + Sort genes by log fold change
#        + Calculate running sum: incremented when gene in set, decremented when not
#        + Maximum of the running sum is enrichment score ES; large ES means that genes in set are toward top of list
#        + Permuting subject labels for signficance
#    + Approach 3: [[http://bioinformatics.oxfordjournals.org/content/23/3/306.long][category t-test]]
#      + Steps
#        + Summarize t (or other) statistic in each set
#        + Test for significance by permuting the subject labels
#        + Much more straight-forward to implement
#    + Approach 4: [[http://bioinformatics.oxfordjournals.org/content/23/8/980.long][Competitive versus self-contained null hypothesis]]
#      + Notes
#        + Competitive null: The genes in the gene set do not have stronger association with the subject condition than other genes. (Approach 1, 2)
#        + Self-contained null: The genes in the gene set do not have any association with the subject condition. (Approach 3)
#        + Probably, self-contained null is closer to actual question of interest
#        + Permuting subjects (rather than genes) is appropriate
#    + Approach 4: [[http://bioinformatics.oxfordjournals.org/content/24/1/78.full.pdf][linear models]]
#      + Notes
#        + Colorectal tumors have good (‘stage II’) or bad (‘stage III’) prognosis
#        + Do genes in the p53 pathway (just one gene set!) show different activity at the two stages?
#        + Linear model incorporates covariates – sex of patient, location of tumor
#    + Approach 5: [[http://bioinformatics.oxfordjournals.org/content/25/1/75.long][pathway topology]]
#      + Notes
#        + Incorporate pathway topology (e.g., interactions between gene products) into signficance testing
#        + Signaling Pathway Impact Analysis
#        + Combined evidence: pathway over-representation PNDE; unusual signaling PPERT (equation 1 of [[http://bioinformatics.oxfordjournals.org/content/25/1/75.long][Tarca et al]])
#    + Approach 6: [[http://www.biomedcentral.com/1471-2105/9/559][de novo discovery]]
#      + Notes
#        + So far: analogous to supervised machine learning, where pathways are known in advance
#        + What about unsupervised discovery?
#        + Weighted correlation network analysis
# 7. Issues with sequence data?
#    + Importantly, all else being equal, long genes receive more reads than short genes
#    + Per-gene P values proportional to gene size
#    + E.g., Young et al., 2010, [[http://bioconductor.org/packages/release/bioc/html/goseq.html][goseq]]
#      + Hypergeometric, weighted by gene size
#      + Substantial differences
#      + Better: read depth??
# 8. Tools For Analysis
#    + [[http://geneontology.org/][GO]]
#    + [[http://proto.informatics.jax.org/prototypes/vlad-1.0.3/][VLAD]]
#    + [[http://pantherdb.org/][Panther]]
#    + [[http://biit.cs.ut.ee/gprofiler/][gProfiler]]
# 9. References
#    + [[https://www.bioconductor.org/help/course-materials/2015/SeattleApr2015/E_GeneSetEnrichment.html][Text Adopted and Adapted from "Gene Set Enrichiment" By Martin Morgan (mtmorgan@fredhutch.org)]]
* 
| *Navigation:*             | *[[https://github.tamu.edu/DigitalBiology/BIOL647_Digital_Biology_2021/wiki][Home]]*                                                                       |
| *Author: [[raramayo@tamu.edu][Rodolfo Aramayo]]* | *License: [[http://creativecommons.org/licenses/by-nc-sa/4.0/][All content produced in this site is licensed by: CC BY-NC-SA 4.0]]* |
